### PR TITLE
[#21] Adds mock tracer.

### LINKED ISSUE #21: Create mock tracer
OpenTracing is a set of interfaces and that means that unit testing your components will relay on specific implementations. For example, I want to assert what the attributes or tags are set in a span as part of my unit testing, what you need is a Tracer that creates a Span which is possible to inspect (exposing all the attributes available to do assertions). Something like:

Some examples:
- https://github.com/opentracing/opentracing-go/blob/master/mocktracer/mockspan.go
- https://github.com/opentracing/opentracing-java/tree/master/opentracing-mock
- https://github.com/opentracing/opentracing-javascript/tree/master/src/mock_tracer

From @yurishkuro in #40 

> Having a mock tracer is useful for two reasons:
> - it allows unit testing of the instrumentation with only the opentracing dependency
> - implementing a mock tracer helps with ironing out the issues in the API itself, like #38, as it serves as a reference implementation

Ping @beberlei @felixfbecker

### PR DESCRIPTION
**Pull request by jcchavezs**
### Short description of what this PR does:
- This PR aims to add the mock tracer for the OpenTracing API.
- Improvements in the `readme.md` are more than welcome.

### Checklist
- [ ] I have made a material change to the repo (functionality, testing, spelling, grammar)
- [x] I have read the [Contribution Guide] and my PR follows them.
- [x] I updated my branch with the master branch.
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation about the functionality in the appropriate .md file
- [x] I have added in line documentation to the code I modified

Ping @yurishkuro @tedsuo @felixfbecker @beberlei

Closes #21

### CODE REVIEW

**Comment by beberlei** [comment]
Why is it named mock tracer? it doesnt really mock or? it really works.

**Comment by jcchavezs** [comment]
Basically following the naming from [other language's libraries](https://github.com/opentracing/opentracing-php/issues/21#issue-260007636). The main advantage of this tracer is that it is inspectable, you can access to all the attributes of a span.

**Comment by beberlei** [comment]
@jcchavezs ok that makes sense then.

A few subjective technical nitpickeries:

1. I would prefer a namespace `OpenTracking\MockTracer` instead of `OpenTracingMock`. The first namespace is usually the vendor and adding another namespace entry to the composer list has a small perf penalty for something that nobody will use in their production code.

2. Naming the classes `Span` and `implenents OTSpan` with use statement rename is a bad practice imho, because it creates a second item in PHPStorm called `Span` and people sometimes make mistakes importing the wrong ones. If the class would be called `MockSpan` this wouldn't happen.

3. you mentioned this is needed for testing, but you have added no tests using the mock tracer, is this intentional and only needed for #45 after merge?

Otherwise +1

**Comment by jcchavezs** [comment]
Thanks for the great feedback, Ben.

1. I am OK to move to the `OpenTracing\MockTracer` namespace but I think the current setup having `OpenTracingMock` in `require-dev` avoid the performance penalty. Am I missing something?
2. I think that is the whole idea about using namespaces, to me having `OpenTracing\MockTracer\MockSpan` looks too much `mock`. Is there any other approach we could sort this out by?
3. My bad, there are some stuff missing here: tests for tracer and also injection and extraction. I added the `WIP` in the title so it is clear.

Again, great feedback!

**Comment by beberlei** [comment]
@jcchavezs you are right about `require-dev`, sorry i didnt see that.

I don't agree with "too much mock", you must see it in context of being `use'd` once at the top so one "Mock" in the namespace name is already not visible anymore. and then if you are using `Span` in your testcode, its not clear if its the interface or the mock span. From a user perspective of this class i prefer to be clearer.

**Comment by ellisv** [comment]
@jcchavezs But if you leave this under `require-dev` other libraries that require this package and want to use those mocks on their test suite will not be able to do so as they will not be autoloadable.

**Comment by jcchavezs** [comment]
Right @ellisv I will change that.

File: src/OpenTracingMock/Tracer.php [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+
+        call_user_func($this->injectors[$format], $spanContext, $carrier);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function extract($format, $carrier)
+    {
+        if (!array_key_exists($format, $this->injectors)) {
```
Comment (masterada) [inline src/OpenTracingMock/Tracer.php]: Shouldn't you check $this->extractors instead?

File: src/OpenTracingMock/Tracer.php [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+    /**
+     * {@inheritdoc}
+     */
+    public function startSpan($operationName, $options = [])
+    {
+        if (!($options instanceof SpanOptions)) {
+            $options = SpanOptions::create($options);
+        }
+
+        if ($options->getReferences()) {
```
Comment (masterada) [inline src/OpenTracingMock/Tracer.php]: I think the condition should be inverted.

File: src/OpenTracingMock/Span.php [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+
+    /**
+     * @var SpanContext
+     */
+    private $context;
+
+    /**
+     * @var array
+     */
+    private $tags;
```
Comment (StymiedSloth) [inline src/OpenTracingMock/Span.php]: This should be set default to an empty array.

This function will error out if not.
    ```public function setTags(array $tags)
    {
        $this->tags = array_merge($this->tags, $tags);
     }```

File: src/OpenTracingMock/Tracer.php (hunk 2 above)
Comment (StymiedSloth) [reply src/OpenTracingMock/Tracer.php]: Could an empty check be an alternative?
```if (empty($options->getReferences())) { ... ```

File: src/OpenTracingMock/Tracer.php [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+
+    public static function create(array $injectors = [], array $extractors = [])
+    {
+        return new self($injectors, $extractors);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function startSpan($operationName, $options = [])
```
Comment (StymiedSloth) [inline src/OpenTracingMock/Tracer.php]: Method should return a span as per the spec

Comment (yurishkuro) [reply src/OpenTracingMock/Tracer.php]: It's a bit concerning seeing comments like this - do we not have a test harness like we have in Python / NodeJS? We even have one for Go, but it's especially important for dynamic languages.

File: src/OpenTracingMock/Tracer.php (hunk 1 above)
Comment (jcchavezs) [reply src/OpenTracingMock/Tracer.php]: Totally, this was a WIP PR so I did not add tests but I just added them now.

File: src/OpenTracingMock/Tracer.php (hunk 4 above)
Comment (jcchavezs) [reply src/OpenTracingMock/Tracer.php]: Totally, this was a WIP so there were many things missing. I just uploaded all the changes and fixed most of the details.

**Comment by jcchavezs** [comment]
I just uploaded all the code and removed the wip tag. Please have a look.

Ping @beberlei @StymiedSloth @masterada @ellisv @yurishkuro

**Comment by jcchavezs** [comment]
Ping @ellisv

File: src/OpenTracing/Mock/MockSpanContext.php [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+    {
+        return array_key_exists($key, $this->items) ? $this->items[$key] : null;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function withBaggageItem($key, $value)
+    {
+        return new self(array_merge($this->items, [$key => $value]));
```
Comment (ellisv) [inline src/OpenTracing/Mock/MockSpanContext.php]: Does this work as expected?

There are 4 constructor arguments, items being the last one. You pass new items argument as first argument only.

File: src/OpenTracing/Mock/MockSpan.php [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+     * {@inheritdoc}
+     */
+    public function setTag($key, $value)
+    {
+        $this->tags[$key] = $value;
+    }
+
+    public function getTags()
+    {
+        return $this->tags;
```
[the remark this answers is not in the document]
Comment (jcchavezs) [reply src/OpenTracing/Mock/MockSpan.php]: Good catch. Added tests for that.

File: src/OpenTracing/Mock/MockSpan.php [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+
+    public function getStartTime()
+    {
+        return $this->startTime;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function finish($finishTime = null, array $logRecords = [])
```
[the remark this answers is not in the document]
Comment (jcchavezs) [reply src/OpenTracing/Mock/MockSpan.php]: True.

File: src/OpenTracing/Mock/MockSpan.php [hunk 8]
```diff
@@ -0,0 +1,167 @@
+<?php
+
+namespace OpenTracing\Mock;
+
+use OpenTracing\ScopeManager;
+use OpenTracing\Span as OTSpan;
```
[the remark this answers is not in the document]
Comment (jcchavezs) [reply src/OpenTracing/Mock/MockSpan.php]: Good catch. This was before we decided to prefix with `Mock` all mock classes.

File: src/OpenTracing/Mock/MockSpanContext.php (hunk 5 above)
Comment (jcchavezs) [reply src/OpenTracing/Mock/MockSpanContext.php]: Great catch! I forgot the test for `MockSpanContext`. I just uploaded it.

**Comment by jcchavezs** [comment]
Please final review @beberlei @StymiedSloth @masterada @ellisv @yurishkuro

**Pull request merged**